README.md
# <img src="https://raw.githubusercontent.com/niladam/tab-limiter/master/icons/tab_50.png"> Tab Limiter

#### A Chrome Extension that limits the number of open, non-pinned tabs to a specified value

This extension was originally created by [necolas](https://github.com/necolas/chrome-tab-limit).

This is an improved version of it, modified to suit my needs and offer some indication as to the tabs opened.

The indicators look like this:

- ![#13B618](https://placehold.it/17/13B618/ffffff?text=+) `if the number of tabs is within a margin of 25% to 50%`
- ![#E77714](https://placehold.it/17/E77714/ffffff?text=+) `if the number of tabs is within a margin of 50% to 75%`
- ![#CD482C](https://placehold.it/17/CD482C/ffffff?text=+) `if the number of tabs is within a margin of 75% to 100%`

## Installation

### Option 1
1. Download the [dist/tab-limiter.crx](dist/tab-limiter.crx)
2. Install.

### Option 2
1. Download or clone this repository
2. Go to `chrome://extensions`
3. Enable developer mode
4. Load the unpacked extension

## Use

Specify the tab limit number by going into options (or click on the tab icon).
By default the limit is set as 20.


CHANGELOG.md
# 1.0

Forked version from [necolas](https://github.com/necolas/chrome-tab-limit)
[fix] Close new tabs that are over the limit.
[add] Options page
[add] Added an option that allows limit specification
[add] Icon changes. New icon comes from [Flaticon](https://www.flaticon.com/authors/roundicons)
[add] Options are saved to Chrome's storage.
[add] The bade background text that holds the tabs number now changes color based on the tabs number
[add] A notification is shown when the limit has been reached, denying the new tab open.

# 0.1.0 (November 18, 2013)

* Initial release. The original code can be found in [necolas's repo](https://github.com/necolas/chrome-tab-limit)


js/options.js
'use strict';

let maxTabs; // Default
let tabsLimit;


// Saves options to chrome.storage.sync.
function save_options() {
  var limit = document.getElementById('limit').value || 20;
  console.log('Limita ' + limit);
  chrome.storage.sync.set({
    tabsLimit: limit
  }, function() {
    // Update status to let user know options were saved.
    var status = document.getElementById('status');
    status.textContent = 'Options saved. Reloading extension..';
    status.style.display = "block";
    setTimeout(function() {
      status.textContent = '';
      status.style.display = "none";
    }, 1500);
    // Let's reload the extension.
    chrome.runtime.reload();
  });
}

// Restores select box and checkbox state using the preferences
// stored in chrome.storage.
function restore_options() {
    chrome.storage.sync.get("tabsLimit", function (data) {
        maxTabs = data.tabsLimit || 20;
        document.getElementById('limit').value = maxTabs;
        // console.log(data);
        // var settingEnabled = data.someSetting;
    });
}
document.addEventListener('DOMContentLoaded', restore_options);
document.getElementById('save').addEventListener('click', save_options);


background.js
// Icons from <div>Icons made by <a href="https://www.flaticon.com/authors/roundicons" title="Roundicons">Roundicons</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a> is licensed by <a href="http://creativecommons.org/licenses/by/3.0/" title="Creative Commons BY 3.0" target="_blank">CC 3.0 BY</a></div>

let isEnabled = true;
var maxTabs;
var tabsCount;
let tabLimit;
let limit;


chrome.storage.sync.get("tabsLimit", function(data) {
    maxTabs = data.tabsLimit;
    // console.log(data);
    console.log(maxTabs);
    // var settingEnabled = data.someSetting;
});

// Notification data
var opt = {
    type: "basic",
    title: "Oh noes!",
    message: "You hit your tabs limit!",
    iconUrl: "icons/tab_128.png",
}

// Updates the badge text and background color.
function updateBadgeText() {
    percent = tabsCount / maxTabs * 100;
    // console.log(percent);
    if (percent >= 25 && percent < 50) {
        chrome.browserAction.setBadgeBackgroundColor({ color: '#13B618' })
    } else if (percent >= 50 && percent < 75) {
        chrome.browserAction.setBadgeBackgroundColor({ color: '#E77714' })
    } else if (percent >= 75 && percent <= 100) {
        chrome.browserAction.setBadgeBackgroundColor({ color: '#CD482C' })
    }
    chrome.browserAction.setBadgeText({ text: "" + tabsCount });
}

// Updates the tab count.
function updateTabsCount() {
    chrome.tabs.query({
        windowType: 'normal',
        pinned: false
    }, function(tabs) {
        tabsCount = tabs.length;
        updateBadgeText();
    });
}

// Handles a new tab and closes it if the limit has been hit.
function handleTabCreated(tab) {
    tabid = tab.id;
    console.log('count:' + tabsCount);
    console.log('max:' + maxTabs);
    if (tabsCount >= maxTabs) {
        // console.log(tab);
        chrome.tabs.remove(tabid);
        chrome.notifications.create(opt);
    } else {
        updateTabsCount();
    }
}

// Handles the tab removed
function handleTabRemoved(tab) {
    updateTabsCount();
}

// This handles when the tab has been updated. Unused for now.
function handleTabUpdated(tab) {
    // updateTabsCount();
}

function init() {
    updateTabsCount();
    chrome.tabs.onCreated.addListener(handleTabCreated);
    chrome.tabs.onRemoved.addListener(handleTabRemoved);
    chrome.tabs.onUpdated.addListener(handleTabUpdated);
}

function teardown() {
    chrome.tabs.onCreated.removeListener(handleTabCreated);
    chrome.tabs.onRemoved.removeListener(handleTabRemoved);
    chrome.tabs.onUpdated.removeListener(handleTabUpdated);
}

chrome.browserAction.onClicked.addListener(function(tab) {
    chrome.tabs.create({ 'url': chrome.extension.getURL('options.html') }, function(tab) {
        // Launching options :)
    });
});

// chrome.browserAction.onClicked.addListener(function (tab) {
//     if (!isEnabled) {
//         init();
//         chrome.tabs.create({'url': chrome.extension.getURL('options.html')}, function(tab) {
//           // Tab opened.
//         });
//         // chrome.browserAction.setIcon({ path: "icons/tab_128.png" });
//     }
//     else {
//         teardown();
//         chrome.tabs.create({'url': chrome.extension.getURL('options.html')}, function(tab) {
//           // Tab opened.
//         });
//         // chrome.browserAction.setIcon({ path: "icons/tab_24.png" });
//         // chrome.browserAction.setBadgeText({ 'text': '' });
//     }

//     isEnabled = !isEnabled;
// });

init();
